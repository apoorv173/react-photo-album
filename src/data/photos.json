[
  {
    "photoId": 1,
    "photoTitle": "quidem molestiae enim",
    "thumbnailUrl": "/fff"
  },
  {
    "photoId": 2,
    "photoTitle": "sunt qui excepturi placeat culpa",
    "thumbnailUrl": "/111"
  }
]