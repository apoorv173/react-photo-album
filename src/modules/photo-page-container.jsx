import React, { Fragment, Suspense, useState, useEffect, lazy } from "react";
import {connect} from "react-redux";
import queryString from 'query-string';
import { fetchPhotos } from "../actions/photo-actions";
import { ModalWindow, Loader, Error } from "../components/";
import { getPhotoDetails } from '../selectors/selectors';

const PhotoList = lazy(() => import("../components/photo-list/photo-list"));
const DataTable = lazy(() => import("../components/data-table/data-table"));
const Pagination = lazy(() => import("../components/pagination/pagination"));

const PhotoPage = ({history, page, match, fetchAllPhotos, album, photosData, loading, error }) =>  {
    
    const [ show, setShow ] = useState(false);
    const [ modalData, setModalData ] = useState({});

    const getQueryParams = queryString.parse(history.location.search);
    const getSizeFromURL = parseInt(getQueryParams["size"], 10) || 20;
    const getPageFromURL = parseInt(getQueryParams["page"], 10) || 1;

    const photoCount = parseInt((photosData ? photosData.length : 0), 10);
    const albumId = parseInt(match.params.albumId, 10);
    
    useEffect(() => {
        fetchAllPhotos(albumId, getPageFromURL, getSizeFromURL);
    }, [match]);

    const onCountChange = (value) => {
		history.push({
            pathname: `/albums/${albumId}/photos`,
			search: `?size=${value}&page=${getPageFromURL}`
		});
    }

    const onPrevClick = () => {
		if(getPageFromURL > 1) {
			const page = getPageFromURL-1;
			history.push({
				pathname: `/albums/${albumId}/photos/`,
				search: `?size=${getSizeFromURL}&page=${page}`
			});
		}
	}

	const onNextClick = () => {
		const page = getPageFromURL+1;
		history.push({
			pathname: `/albums/${albumId}/photos`,
			search: `?size=${getSizeFromURL}&page=${page}`
		});
    }
    
    const onPhotoClick = (obj) => {
        setShow(true);
        setModalData(obj);
    }

    const closeModalHandler = () => {
        setShow(false);
        setModalData({});
    }

	const renderModalBody = () => {
        return (
            <Fragment>
                <div className="modal-right">
                    <img alt={modalData.title} className="thumbnail-img" src={modalData.url}/>
                </div>
                <div className="modal-left">
                    <h1 className="modal-title">{modalData.photoTitle}</h1>
                    <h2>By <span>{`   `}</span>{modalData.username}</h2>
                    <p className="modal-desc">Album: <span>{`   `}</span>{modalData.albumTitle}</p>
                </div>
            </Fragment>
        );
	}

	const renderModalWindow = () => {
		return (
			<ModalWindow
				className="modal"
				show={show}
				onClose={closeModalHandler}
			>
				{renderModalBody()}
			</ModalWindow>
		);
    }
    switch(true) {
        case (loading === true): 
            return <Loader />;

        case (error && error !== ''):
            return <Error error={error} />;

        case (photosData && photosData.length >= 0 ):
            return (
                <Suspense fallback={<Loader />} >
                    <Pagination 
                        prevDisabled={(getPageFromURL <= 1)}
                        nextDisabled={photoCount < getSizeFromURL}
                        onPrevClick={onPrevClick}
                        onNextClick={onNextClick}
                    />
                    <DataTable
                        rowCount={getSizeFromURL}
                        onCountChange={onCountChange}
                    />
                    <PhotoList 
                        album={album}       
                        photos={photosData}
                        page={page}
                        onPhotoClick={onPhotoClick}
                    />
                    {(show && modalData) && renderModalWindow()}

                </Suspense>
            )
            default: 
                return null;
    }
}

const mapStateToProps = state => {
    return {
        photosData: getPhotoDetails(state),
        error: state.photos.error,
        loading: state.photos.loading
        
    }
};

const mapDispatchToProps = dispatch => {
    return {
        fetchAllPhotos: (albumId, page, limit) => {
          dispatch(fetchPhotos(albumId, page, limit));
        }
      }
};

export default connect(mapStateToProps, mapDispatchToProps)(PhotoPage);
