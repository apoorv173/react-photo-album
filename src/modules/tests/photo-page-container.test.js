import React, { lazy } from "react";
import {MemoryRouter} from "react-router-dom";
import { shallow, mount } from "enzyme";
import { PhotoPage } from "../photo-page-container";
import photos from '../../data/more-photos.json';
import { history } from "../../configureStore";

const PhotoList = lazy(() => import("../components/photo-list/photo-list"));
const DataTable = lazy(() => import("../components/data-table/data-table"));
const Pagination = lazy(() => import("../../components/pagination/pagination"));

describe("Photo List Component", () => {

    const onNextClick = jest.fn();
    const onPrevClick = jest.fn();

    it("Should render the loader component only with its snapshot", () => {
        const tree = shallow(<MemoryRouter><PhotoPage loading={true} /></MemoryRouter>)
        expect(tree).toMatchSnapshot();
    });

    it("Should render the error component only with its snapshot", () => {
        const tree = shallow(<MemoryRouter><PhotoPage error="no results found" /></MemoryRouter>)
        expect(tree).toMatchSnapshot();
    });

    it("Should return null if albums doesnt exist, and no eror is there", () => {
        const tree = shallow(<MemoryRouter><PhotoPage loading={false} error="" albumsData={[]} /></MemoryRouter>)
        expect(tree).toMatchSnapshot();
    });
    
    it('rendered lazily', async()=> {
        const fetchAllPhotos = jest.fn();
        const match = {
            params: {
                albumId: '1'
            }
        }
        const wrapper = mount(
            <MemoryRouter>
                <PhotoPage  photosData={photos} match={match} loading={false} error='' history={history} fetchAllPhotos={fetchAllPhotos} >
                </PhotoPage>
            </MemoryRouter>
        );
    
        await PhotoList;
        await DataTable;
        await Pagination;

        wrapper.setProps({ photosData: photos, loading: false });
        wrapper.find('.next-btn').simulate('click');
        expect(history.location.pathname).toEqual('/albums/1/photos');
        expect(history.location.search).toEqual('?size=20&page=2');
      })

});