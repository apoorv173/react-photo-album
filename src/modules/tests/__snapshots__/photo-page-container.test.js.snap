// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album List Component Should render the error component only with its snapshot 1`] = `ShallowWrapper {}`;

exports[`Album List Component Should render the loader component only with its snapshot 1`] = `ShallowWrapper {}`;

exports[`Album List Component Should return null if albums doesnt exist, and no eror is there 1`] = `ShallowWrapper {}`;

exports[`Photo List Component Should render the error component only with its snapshot 1`] = `ShallowWrapper {}`;

exports[`Photo List Component Should render the loader component only with its snapshot 1`] = `ShallowWrapper {}`;

exports[`Photo List Component Should return null if albums doesnt exist, and no eror is there 1`] = `ShallowWrapper {}`;
